xbmc/games/addons/cheevos/GameClientCheevos.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+ *  Copyright (C) 2020-2021 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "GameClientCheevos.h"
+
+#include "addons/kodi-dev-kit/include/kodi/c-api/addon-instance/game.h"
+#include "cores/RetroPlayer/cheevos/RConsoleIDs.h"
+#include "games/addons/GameClient.h"
+
+using namespace KODI;
+using namespace GAME;
+
+CGameClientCheevos::CGameClientCheevos(CGameClient& gameClient, AddonInstance_Game& addonStruct)
+  : m_gameClient(gameClient), m_struct(addonStruct)
+{
+}
+
+bool CGameClientCheevos::RCGenerateHashFromFile(std::string& hash,
+                                                RETRO::RConsoleID consoleID,
+                                                const std::string& filePath)
+{
+  char* _hash = nullptr;
+  GAME_ERROR error = GAME_ERROR_NO_ERROR;
+
+  try
+  {
+    m_gameClient.LogError(
+        error = m_struct.toAddon->RCGenerateHashFromFile(
+            &m_struct, &_hash, static_cast<unsigned int>(consoleID), filePath.c_str()),
+        "RCGenerateHashFromFile()");
+  }
+  catch (...)
+  {
+    m_gameClient.LogException("RCGetGameIDUrl()");
+  }
+
+  if (_hash)
+  {
+    hash = _hash;
+    free(_hash);
```

Comment:
I just caught this. `free()` must be called within the add-on. I discovered long ago when working on RetroPlayer that freeing memory across a DLL boundary causes bad things to happen. This means we either need a `FreeRC***()` that calls free(), or we use a static buffer of a fixed size.

Reply:
> I just caught this. free() must be called within the add-on.

Shouldn't really matter if it's done using the "C" functions "malloc" or "strdup" on the addon, it should actually always work in Kodi using "free" and carried out purely within the add-on API headers.
Of course, it could look different if the memory came from e.g. RetroPlayer itself, since you can never be sure on which basis it created it.

----------------------------------

In my API rework using a sandbox, I also found it difficult to manage easily with free calls to addons. The dilemma there is that Kodi uses a different memory than the addon described at the beginning. The auto generation from me then makes something like a new malloc to Kodi.

Here is a mini example of how something is done in it.

```cpp
// Kodi's place call:
char* a = nullptr;
const char b = "example";
void GetSomething(char** a, const char* b)
{
    struct SOME_SHARED_MEM {
    int what_to_do;
    char some_available_shared_mem[HIS_SIZE];
  } c_on_kodi;
  strcpy(c_on_kodi.some_available_shared_mem, b);
  STOP_KODI_AND_INFORM_ADDON();
  ...
    ---------------------------------------------------------
    SOME_SHARED_MEM c_on_addon; // Same pointer as c_on_kodi
    char* a = nullptr;
    GetSomething(a, c_on_addon.some_available_shared_mem)
    {
      *a = strdup("SOMETHING");
    }
    strcpy(c_on_addon.some_available_shared_mem, a);
    free(a);
    STOP_ADDON_AND_INFORM_KODI();
    ---------------------------------------------------------
  ...
  // Now back on Kodi in function "GetSomething"
  *a = strdup(c_on_kodi.some_available_shared_mem);
}

// Somewhere later
free(a);
```
In the right code much more complex only here as a mini.

So the caller of "GetSomething" has to make his own free, since it can never pass the right pointer in a callback to the addon.

Fixed memory is also very easy with the Sandbox, but I always find it a bit dangerous because you can never be 100% sure that it won't exceed your size. It also permanently reserves unnecessary memory.

Reply:
So do you propose we leave it like this to merge, and re-address it later when the sandbox lands?

Reply:
Within Sandbox there would be almost nothing to change here.
See here as an example from the audiodecoder system which has also been using `free` for strings for a while:

- in Sandbox: https://github.com/AlwinEsch/kodi/blob/new-interface-2/xbmc/addons/AudioDecoder.cpp#L118-L219
- In current Kodi master: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AudioDecoder.cpp#L130-L233

The only difference is:
```cpp
   const auto ifc = m_ifc->kodi_addoninstance_audiodecoder_h;
   bool ret = ifc->kodi_addon_audiodecoder_read_tag_v1(m_instance, fileName.c_str(), &ifcTag);
```

Here is part of the automatically generated code on the Kodi side, with which it is added to the addon. There on line 232 is the direct call (as currently) if sandbox were deactivated:
https://github.com/AlwinEsch/kodi/blob/new-interface-2/xbmc/addons/interface/api/addon-instance/audiodecoder.cpp#L205-L234